packages/back-end/src/models/DataSourceModel.ts

```diff
@@ -205,27 +211,69 @@ export async function createDataSource(
   return toInterface(model);
 }
 
+// Add any missing exposure query ids and validate any new, changed, or previously errored queries
+export async function validateExposureQueriesAndAddMissingIds(
+  datasource: DataSourceInterface,
+  updates: Partial<DataSourceInterface>
+): Promise<Partial<DataSourceInterface>> {
+  const updatesCopy = cloneDeep(updates);
+  if (updatesCopy.settings?.queries?.exposure) {
+    await Promise.all(
+      updatesCopy.settings.queries.exposure.map(async (exposure) => {
+        let checkValidity = false;
+        if (!exposure.id) {
+          exposure.id = uniqid("exq_");
+          checkValidity = true;
+        } else {
+          const existingQuery = datasource.settings.queries?.exposure?.find(
+            (q) => q.id == exposure.id
+          );
+          if (
+            !existingQuery ||
+            existingQuery.query != exposure.query ||
+            existingQuery.error
```

Comment:
There are other changes that could affect if there's an error or not. For example, changing the `hasNameColumns` or `dimensions` setting changes which columns are required to be returned.